app/src/tunnel/kotlin/tunnel/Blockade.kt

```diff
@@ -36,31 +51,97 @@ internal class Blockade(
                             )
                         } else {
                             ktx.e("memory limit reached, skipping ruleset", nextFilter,
-                                    doGetMemoryLimit(), ruleset.size)
+                                       doGetMemoryLimit(), ruleset.size)
                         }
                     }
                 },
                 failure = {
                     ktx.e("could not load first ruleset", filters.first(), it)
                 }
         )
+        if (order == 3) {
+            var orignalrulesize = ruleset.size
+            //TODO add button that says "Octomize" in advanced settings. have that button switch the below var octomize on/off.
+            // Having octomize = true reduces list by about 20% and thus saves memory but takes a little longer to update.
+            // update time is an issue due to the fact that blokada seems to completely re-download each list and do the update process four times.
+            // faedrak said he thinks the issue is: the function gets called asynchron based on changes of a variable but has no check for other already running instances."
+            var octomize = true
+            if ( wildcardRuleset.size==0)
+                octomize = false
+            var counter = orignalrulesize - ruleset.size // because some were removed from whitelist
```

Comment:
That will mess with the percentages. Because some URLS that were listed in whitelist, where removed from blacklist, while addressing https://github.com/blokadaorg/blokada/issues/373.

android.util.Log.d("skipped", ""+ (orignalrulesize-counter)+ " " + orignalrulesize + " " + ruleset.size + "  percent finished is " +((counter*100)/orignalrulesize) +  "% percent saved is " + (((orignalrulesize - ruleset.size)*100)/counter)+ "% " + orignalrulesize +" " + ruleset.size + " " + counter + " " + iit + " memory is " + Runtime.getRuntime().freeMemory() + " "  )

Reply:
In line 63 you are setting `orignalrulesize` to `ruleset.size`. This value is not changed between line 63 and line 71 so you are effectively subtracting `ruleset.size` from itself. The result will always be 0.

Reply:
Ah, I now the problem. During the code cleanup somehow I moved the following line 
ruleset.removeAll(whitelistRuleset)

this line was supposed to in between lines 63 and 71. I inserted what I believe to be corrected code below. 

            var orignalrulesize = ruleset.size
            //TODO add button that says "optimize" in advanced settings. have that button switch the below var octomize on/off.
            // Having optimize = true reduces list by about 20% and thus saves memory but takes a little longer to update.
            // update time is an issue due to the fact that blokada seems to completely re-download each list and do the update process four times.
            // faedrak said he thinks the issue is: the function gets called asynchron based on changes of a variable but has no check for other already running instances."
            var optimize = true
            if ( wildcardRuleset.size==0)
                optimize = false
            ruleset.removeAll(whitelistRuleset)
            ruleset.removeAll(wildcardRuleset)
            var counter = orignalrulesize - ruleset.size // because some were removed from whitelist
            val arraylist = ruleset.toString().split(",")//", ", "[", "]")

Reply:
You are using counter only to create less log entrys so it's not important what the start value is.

Reply:
Just thought it would be a good idea to somehow show the user how much the list was optimized so they can see the difference it makes. Right now, the optimization process can make the "loading" aka "uploading" process take a significant amount of time, because of the bug that causes the list to be completely re-downloaded, then everything uploaded, then the process starts again several times. The reason it now can take significantly longer is because the optimization process takes a little time and because the process repeats several times, it is now a noticeable burden. I was thinking of showing the user how much the list was optimized. To do that, you really need to know the list size before any optimization started, and the list size when completed. I just haven't yet figured out how to show the user how much the list was optimized yet.

Reply:
I understand what you are trying to do but it has nothing to do with the fact that it would make more sence to init the `counter `with 0.